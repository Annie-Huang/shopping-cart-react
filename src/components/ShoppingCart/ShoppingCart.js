import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import ShoppingCartService from '../../lib/ShoppingCartService';

export class ShoppingCart extends Component {

    render() {
        const children = this.props.basket.cartItems.map((cartItem, index) =>
            <div className='row' key={index}>
                <div className='col-md-2'>{cartItem.product.name}</div>
                <div className='col-md-2'>Unit Price: ${cartItem.product.price}</div>
                <div className='col-md-2'>Quantity: {cartItem.quantity}</div>
                <div className='col-md-2'>Subtotal: ${cartItem.subTotal}</div>
                {cartItem.discount !== 0 && <div className='col-md-2 text-danger'>Discount: ${cartItem.discount}</div>}
            </div>
        );

        return (
            this.props.showBasket &&
            <div className="card">
                <div className="card-header bg-info text-white">
                    <h3 className="card-title">Your Basket:</h3>
                </div>
                <div className="card-body">
                    {children}
                    <br/>
                    <div className='row'>
                        <div className='col-md-12'>
                        <h3 className="card-title">Total: ${this.props.basket.total}</h3>
                        </div>
                    </div>
                </div>
                <div className="card-footer">
                    <button type="button" className="btn btn-lg btn-primary">Checkout</button>
                </div>
            </div>
        );
    }
}

ShoppingCart.propTypes = {
    basket: PropTypes.object.isRequired,
    showBasket: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
    basket: (state.user.id ?
        ShoppingCartService.calculateCart(state.cartItems, state.user.pricingRules) :
        {cartItems: state.cartItems, total: 0}
    ),
    showBasket: state.cartItems.length > 0
});

export default connect(mapStateToProps, null)(ShoppingCart);
