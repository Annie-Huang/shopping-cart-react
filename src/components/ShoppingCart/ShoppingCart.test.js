import {shallow} from "enzyme/build";
import {ShoppingCart} from "./ShoppingCart";
import React from "react";


describe('ShoppingCart presentation', () => {
    const cartItems = require('../../resources/fixtures/cart-items.json');
    cartItems[0].subTotal = 269.99;
    cartItems[0].discount = 0;
    cartItems[1].subTotal = 619.98;
    cartItems[1].discount = 26;

    const basket = {
        cartItems: cartItems,
        total: 889.97
    };
    const showBasket = true;
    const props = {
        basket,
        showBasket
    };

    it('should render ShoppingCart component when no user selected', () => {
        expect(shallow(<ShoppingCart {...props} />)).toMatchSnapshot();
    });

});
