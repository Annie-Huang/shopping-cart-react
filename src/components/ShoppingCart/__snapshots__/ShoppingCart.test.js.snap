// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart presentation should render ShoppingCart component when no user selected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShoppingCart
    basket={
      Object {
        "cartItems": Array [
          Object {
            "discount": 0,
            "product": Object {
              "id": "classic",
              "name": "Classic Ad",
              "price": 269.99,
            },
            "quantity": 1,
            "subTotal": 269.99,
          },
          Object {
            "discount": 26,
            "product": Object {
              "id": "standout",
              "name": "Standout Ad",
              "price": 322.99,
            },
            "quantity": 2,
            "subTotal": 619.98,
          },
        ],
        "total": 889.97,
      }
    }
    showBasket={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="card-header bg-info text-white"
        >
          <h3
            className="card-title"
          >
            Your Basket:
          </h3>
        </div>,
        <div
          className="card-body"
        >
          <div
            className="row"
          >
            <div
              className="col-md-2"
            >
              Classic Ad
            </div>
            <div
              className="col-md-2"
            >
              Unit Price: $
              269.99
            </div>
            <div
              className="col-md-2"
            >
              Quantity: 
              1
            </div>
            <div
              className="col-md-2"
            >
              Subtotal: $
              269.99
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="col-md-2"
            >
              Standout Ad
            </div>
            <div
              className="col-md-2"
            >
              Unit Price: $
              322.99
            </div>
            <div
              className="col-md-2"
            >
              Quantity: 
              2
            </div>
            <div
              className="col-md-2"
            >
              Subtotal: $
              619.98
            </div>
            <div
              className="col-md-2 text-danger"
            >
              Discount: $
              26
            </div>
          </div>
          <br />
          <div
            className="row"
          >
            <div
              className="col-md-12"
            >
              <h3
                className="card-title"
              >
                Total: $
                889.97
              </h3>
            </div>
          </div>
        </div>,
        <div
          className="card-footer"
        >
          <button
            className="btn btn-lg btn-primary"
            type="button"
          >
            Checkout
          </button>
        </div>,
      ],
      "className": "card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h3
            className="card-title"
          >
            Your Basket:
          </h3>,
          "className": "card-header bg-info text-white",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Your Basket:",
            "className": "card-title",
          },
          "ref": null,
          "rendered": "Your Basket:",
          "type": "h3",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [
              <div
                className="row"
              >
                <div
                  className="col-md-2"
                >
                  Classic Ad
                </div>
                <div
                  className="col-md-2"
                >
                  Unit Price: $
                  269.99
                </div>
                <div
                  className="col-md-2"
                >
                  Quantity: 
                  1
                </div>
                <div
                  className="col-md-2"
                >
                  Subtotal: $
                  269.99
                </div>
              </div>,
              <div
                className="row"
              >
                <div
                  className="col-md-2"
                >
                  Standout Ad
                </div>
                <div
                  className="col-md-2"
                >
                  Unit Price: $
                  322.99
                </div>
                <div
                  className="col-md-2"
                >
                  Quantity: 
                  2
                </div>
                <div
                  className="col-md-2"
                >
                  Subtotal: $
                  619.98
                </div>
                <div
                  className="col-md-2 text-danger"
                >
                  Discount: $
                  26
                </div>
              </div>,
            ],
            <br />,
            <div
              className="row"
            >
              <div
                className="col-md-12"
              >
                <h3
                  className="card-title"
                >
                  Total: $
                  889.97
                </h3>
              </div>
            </div>,
          ],
          "className": "card-body",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="col-md-2"
                >
                  Classic Ad
                </div>,
                <div
                  className="col-md-2"
                >
                  Unit Price: $
                  269.99
                </div>,
                <div
                  className="col-md-2"
                >
                  Quantity: 
                  1
                </div>,
                <div
                  className="col-md-2"
                >
                  Subtotal: $
                  269.99
                </div>,
                false,
              ],
              "className": "row",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Classic Ad",
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": "Classic Ad",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Unit Price: $",
                    269.99,
                  ],
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": Array [
                  "Unit Price: $",
                  269.99,
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Quantity: ",
                    1,
                  ],
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": Array [
                  "Quantity: ",
                  1,
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Subtotal: $",
                    269.99,
                  ],
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": Array [
                  "Subtotal: $",
                  269.99,
                ],
                "type": "div",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="col-md-2"
                >
                  Standout Ad
                </div>,
                <div
                  className="col-md-2"
                >
                  Unit Price: $
                  322.99
                </div>,
                <div
                  className="col-md-2"
                >
                  Quantity: 
                  2
                </div>,
                <div
                  className="col-md-2"
                >
                  Subtotal: $
                  619.98
                </div>,
                <div
                  className="col-md-2 text-danger"
                >
                  Discount: $
                  26
                </div>,
              ],
              "className": "row",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Standout Ad",
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": "Standout Ad",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Unit Price: $",
                    322.99,
                  ],
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": Array [
                  "Unit Price: $",
                  322.99,
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Quantity: ",
                    2,
                  ],
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": Array [
                  "Quantity: ",
                  2,
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Subtotal: $",
                    619.98,
                  ],
                  "className": "col-md-2",
                },
                "ref": null,
                "rendered": Array [
                  "Subtotal: $",
                  619.98,
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Discount: $",
                    26,
                  ],
                  "className": "col-md-2 text-danger",
                },
                "ref": null,
                "rendered": Array [
                  "Discount: $",
                  26,
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="col-md-12"
              >
                <h3
                  className="card-title"
                >
                  Total: $
                  889.97
                </h3>
              </div>,
              "className": "row",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <h3
                  className="card-title"
                >
                  Total: $
                  889.97
                </h3>,
                "className": "col-md-12",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Total: $",
                    889.97,
                  ],
                  "className": "card-title",
                },
                "ref": null,
                "rendered": Array [
                  "Total: $",
                  889.97,
                ],
                "type": "h3",
              },
              "type": "div",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="btn btn-lg btn-primary"
            type="button"
          >
            Checkout
          </button>,
          "className": "card-footer",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Checkout",
            "className": "btn btn-lg btn-primary",
            "type": "button",
          },
          "ref": null,
          "rendered": "Checkout",
          "type": "button",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="card-header bg-info text-white"
          >
            <h3
              className="card-title"
            >
              Your Basket:
            </h3>
          </div>,
          <div
            className="card-body"
          >
            <div
              className="row"
            >
              <div
                className="col-md-2"
              >
                Classic Ad
              </div>
              <div
                className="col-md-2"
              >
                Unit Price: $
                269.99
              </div>
              <div
                className="col-md-2"
              >
                Quantity: 
                1
              </div>
              <div
                className="col-md-2"
              >
                Subtotal: $
                269.99
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="col-md-2"
              >
                Standout Ad
              </div>
              <div
                className="col-md-2"
              >
                Unit Price: $
                322.99
              </div>
              <div
                className="col-md-2"
              >
                Quantity: 
                2
              </div>
              <div
                className="col-md-2"
              >
                Subtotal: $
                619.98
              </div>
              <div
                className="col-md-2 text-danger"
              >
                Discount: $
                26
              </div>
            </div>
            <br />
            <div
              className="row"
            >
              <div
                className="col-md-12"
              >
                <h3
                  className="card-title"
                >
                  Total: $
                  889.97
                </h3>
              </div>
            </div>
          </div>,
          <div
            className="card-footer"
          >
            <button
              className="btn btn-lg btn-primary"
              type="button"
            >
              Checkout
            </button>
          </div>,
        ],
        "className": "card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h3
              className="card-title"
            >
              Your Basket:
            </h3>,
            "className": "card-header bg-info text-white",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Your Basket:",
              "className": "card-title",
            },
            "ref": null,
            "rendered": "Your Basket:",
            "type": "h3",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [
                <div
                  className="row"
                >
                  <div
                    className="col-md-2"
                  >
                    Classic Ad
                  </div>
                  <div
                    className="col-md-2"
                  >
                    Unit Price: $
                    269.99
                  </div>
                  <div
                    className="col-md-2"
                  >
                    Quantity: 
                    1
                  </div>
                  <div
                    className="col-md-2"
                  >
                    Subtotal: $
                    269.99
                  </div>
                </div>,
                <div
                  className="row"
                >
                  <div
                    className="col-md-2"
                  >
                    Standout Ad
                  </div>
                  <div
                    className="col-md-2"
                  >
                    Unit Price: $
                    322.99
                  </div>
                  <div
                    className="col-md-2"
                  >
                    Quantity: 
                    2
                  </div>
                  <div
                    className="col-md-2"
                  >
                    Subtotal: $
                    619.98
                  </div>
                  <div
                    className="col-md-2 text-danger"
                  >
                    Discount: $
                    26
                  </div>
                </div>,
              ],
              <br />,
              <div
                className="row"
              >
                <div
                  className="col-md-12"
                >
                  <h3
                    className="card-title"
                  >
                    Total: $
                    889.97
                  </h3>
                </div>
              </div>,
            ],
            "className": "card-body",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="col-md-2"
                  >
                    Classic Ad
                  </div>,
                  <div
                    className="col-md-2"
                  >
                    Unit Price: $
                    269.99
                  </div>,
                  <div
                    className="col-md-2"
                  >
                    Quantity: 
                    1
                  </div>,
                  <div
                    className="col-md-2"
                  >
                    Subtotal: $
                    269.99
                  </div>,
                  false,
                ],
                "className": "row",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Classic Ad",
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": "Classic Ad",
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Unit Price: $",
                      269.99,
                    ],
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Unit Price: $",
                    269.99,
                  ],
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Quantity: ",
                      1,
                    ],
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Quantity: ",
                    1,
                  ],
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Subtotal: $",
                      269.99,
                    ],
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Subtotal: $",
                    269.99,
                  ],
                  "type": "div",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="col-md-2"
                  >
                    Standout Ad
                  </div>,
                  <div
                    className="col-md-2"
                  >
                    Unit Price: $
                    322.99
                  </div>,
                  <div
                    className="col-md-2"
                  >
                    Quantity: 
                    2
                  </div>,
                  <div
                    className="col-md-2"
                  >
                    Subtotal: $
                    619.98
                  </div>,
                  <div
                    className="col-md-2 text-danger"
                  >
                    Discount: $
                    26
                  </div>,
                ],
                "className": "row",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Standout Ad",
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": "Standout Ad",
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Unit Price: $",
                      322.99,
                    ],
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Unit Price: $",
                    322.99,
                  ],
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Quantity: ",
                      2,
                    ],
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Quantity: ",
                    2,
                  ],
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Subtotal: $",
                      619.98,
                    ],
                    "className": "col-md-2",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Subtotal: $",
                    619.98,
                  ],
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Discount: $",
                      26,
                    ],
                    "className": "col-md-2 text-danger",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Discount: $",
                    26,
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="col-md-12"
                >
                  <h3
                    className="card-title"
                  >
                    Total: $
                    889.97
                  </h3>
                </div>,
                "className": "row",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <h3
                    className="card-title"
                  >
                    Total: $
                    889.97
                  </h3>,
                  "className": "col-md-12",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Total: $",
                      889.97,
                    ],
                    "className": "card-title",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Total: $",
                    889.97,
                  ],
                  "type": "h3",
                },
                "type": "div",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="btn btn-lg btn-primary"
              type="button"
            >
              Checkout
            </button>,
            "className": "card-footer",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Checkout",
              "className": "btn btn-lg btn-primary",
              "type": "button",
            },
            "ref": null,
            "rendered": "Checkout",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
